File: ToDo.java
import java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;

class Task {
    private String name;
    private String description;
    private String dueDate;
    private boolean completed;

    // Task Constructor
    public Task(String name, String description, String dueDate) {
        this.name = name;
        this.description = description;
        this.dueDate = dueDate;
        this.completed = false;
    }

    public String getName() {
        return name;
    }

    public String getDueDate() {
        return dueDate;
    }

    public boolean isCompleted() {
        return completed;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public void setDueDate(String dueDate) {
        this.dueDate = dueDate;
    }

    public void markCompleted() {
        this.completed = true;
    }

    public void displayTask() {
        System.out.println("\t   Task Name: " + name);
        System.out.println("\t   Description: " + description);
        System.out.println("\t   Due Date: " + dueDate);
        System.out.println("\t   Status: " + (completed ? "Completed" : "Incomplete"));
        System.out.println("\t   ------------------------------");
    }
}

class ToDoListApp {
    private List<Task> taskList;

    public ToDoListApp() {
        taskList = new ArrayList<>();
    }

    private void sortTasksByDueDate() {
        taskList.sort(Comparator.comparing(Task::getDueDate));
    }

    public void addTask(String name, String description, String dueDate) {
        taskList.add(new Task(name, description, dueDate));
        System.out.println("\t   Task added successfully!");
    }

    public void displayTasks() {
        if (taskList.isEmpty()) {
            System.out.println("\t   Task list is empty.");
            return;
        }
        sortTasksByDueDate();
        for (Task task : taskList) {
            task.displayTask();
        }
    }

    public void markTaskCompleted(String taskName) {
        for (Task task : taskList) {
            if (task.getName().equalsIgnoreCase(taskName)) {
                task.markCompleted();
                System.out.println("\t   Task '" + taskName + "' marked as completed.");
                return;
            }
        }
        System.out.println("\t   Task '" + taskName + "' not found.");
    }

    public void deleteTask(String taskName) {
        Iterator<Task> iterator = taskList.iterator();
        while (iterator.hasNext()) {
            if (iterator.next().getName().equalsIgnoreCase(taskName)) {
                iterator.remove();
                System.out.println("\t   Task '" + taskName + "' deleted successfully!");
                return;
            }
        }
        System.out.println("\t   Task '" + taskName + "' not found.");
    }

    public void editTask(String taskName, String newDescription, String newDueDate) {
        for (Task task : taskList) {
            if (task.getName().equalsIgnoreCase(taskName)) {
                task.setDescription(newDescription);
                task.setDueDate(newDueDate);
                System.out.println("\t   Task updated successfully!");
                return;
            }
        }
        System.out.println("\t   Task '" + taskName + "' not found.");
    }

    private boolean isDueSoon(String currentDate, String dueDate) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
        try {
            Date current = sdf.parse(currentDate);
            Date due = sdf.parse(dueDate);

            long diff = (due.getTime() - current.getTime()) / (1000 * 60 * 60 * 24);
            return diff == 0 || diff == 1; // Due today or tomorrow
        } catch (ParseException e) {
            System.out.println("\t   Error parsing dates.");
        }
        return false;
    }

    public void showReminders(String currentDate) {
        boolean found = false;
        System.out.println("\t   Upcoming tasks due soon:");
        for (Task task : taskList) {
            if (!task.isCompleted() && isDueSoon(currentDate, task.getDueDate())) {
                task.displayTask();
                found = true;
            }
        }
        if (!found) {
            System.out.println("\t   No tasks are due within the next day.");
        }
    }
}

public class ToDoListAppMain {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        ToDoListApp toDoListApp = new ToDoListApp();
        int choice;

        do {
            System.out.println("=============================================");
            System.out.println("\t\t    To-Do List Application");
            System.out.println("=============================================");
            System.out.println("1. Add a new task");
            System.out.println("2. Display all tasks");
            System.out.println("3. Mark task as completed");
            System.out.println("4. Delete a task");
            System.out.println("5. Edit a task");
            System.out.println("6. Show reminders");
            System.out.println("7. Exit");
            System.out.print("Enter your choice: ");
            choice = sc.nextInt();
            sc.nextLine(); // Consume newline

            switch (choice) {
                case 1:
                    System.out.print("Enter task name: ");
                    String name = sc.nextLine();
                    System.out.print("Enter task description: ");
                    String description = sc.nextLine();
                    System.out.print("Enter due date (YYYY-MM-DD): ");
                    String dueDate = sc.nextLine();
                    toDoListApp.addTask(name, description, dueDate);
                    break;

                case 2:
                    toDoListApp.displayTasks();
                    break;

                case 3:
                    System.out.print("Enter the name of the task to mark as completed: ");
                    String completedTask = sc.nextLine();
                    toDoListApp.markTaskCompleted(completedTask);
                    break;

                case 4:
                    System.out.print("Enter the name of the task to delete: ");
                    String deleteTask = sc.nextLine();
                    toDoListApp.deleteTask(deleteTask);
                    break;

                case 5:
                    System.out.print("Enter the name of the task to edit: ");
                    String editTask = sc.nextLine();
                    System.out.print("Enter new description: ");
                    String newDescription = sc.nextLine();
                    System.out.print("Enter new due date (YYYY-MM-DD): ");
                    String newDueDate = sc.nextLine();
                    toDoListApp.editTask(editTask, newDescription, newDueDate);
                    break;

                case 6:
                    System.out.print("Enter the current date (YYYY-MM-DD): ");
                    String currentDate = sc.nextLine();
                    toDoListApp.showReminders(currentDate);
                    break;

                case 7:
                    System.out.println("Exiting application. Goodbye!");
                    break;

                default:
                    System.out.println("Invalid choice. Please try again.");
                    break;
            }
        } while (choice != 7);

        sc.close();
    }
}

File: ToDo.cpp
#include <bits/stdc++.h>
#include <conio.h>
using namespace std;

// Task class to represent each task
class Task
{
public:
    string name;
    string description;
    string dueDate;
    bool completed;

    // Task Constructor
    Task(string name, string description, string dueDate)
    {
        this->name = name;
        this->description = description;
        this->dueDate = dueDate;
        this->completed = false;
    }

    // Display task details
    void displayTask()
    {
        cout << "\t\t\t\t\t   Task Name: " << name << "\n";
        cout << "\t\t\t\t\t   Description: " << description << "\n";
        cout << "\t\t\t\t\t   Due Date: " << dueDate << "\n";
        cout << "\t\t\t\t\t   Status: " << (completed ? "Completed" : "Incomplete") << "\n";
        cout << "\t\t\t\t\t   ------------------------------\n";
    }
};

// Main class to manage tasks
class ToDoListApp
{
public:
    vector<Task> taskList;

    void sortTasksByDueDate()
    {
        sort(taskList.begin(), taskList.end(), [](const Task &a, const Task &b)
             { return a.dueDate < b.dueDate; });
    }

    // Method to add a new task
    void addTask(string name, string description, string dueDate)
    {
        Task newTask(name, description, dueDate);
        taskList.push_back(newTask);
        cout << "\t\t\t\t\t   Task added successfully!\n";
    }

    // Display all tasks sorted by due date
    void displayTasks()
    {
        if (taskList.empty())
        {
            cout << "\t\t\t\t\t   Task list is empty.\n";
            return;
        }
        sortTasksByDueDate();
        for (auto &task : taskList)
        {
            task.displayTask();
        }
    }

    // Mark a task as completed
    void markTaskCompleted(const string &taskName)
    {
        for (auto &task : taskList)
        {
            if (task.name == taskName)
            {
                task.completed = true;
                cout << "\t\t\t\t\t   Task '" << taskName << "' marked as completed.\n";
                return;
            }
        }
        cout << "\t\t\t\t\t   Task '" << taskName << "' not found.\n";
    }

    // Delete a task
    void deleteTask(string taskName)
    {
        auto it = remove_if(taskList.begin(), taskList.end(), [&taskName](const Task &task)
                            { return task.name == taskName; });

        if (it != taskList.end())
        {
            taskList.erase(it, taskList.end());
            cout << "\t\t\t\t\t   Task '" << taskName << "' deleted successfully!\n";
        }
        else
        {
            cout << "\t\t\t\t\t   Task '" << taskName << "' not found.\n";
        }
    }

    // Edit task if needed
    void editTask(string taskName)
    {
        for (auto &task : taskList)
        {
            if (task.name == taskName)
            {
                cout << "\t\t\t\t\t   Editing Task: " << taskName << "\n";
                cout << "\t\t\t\t\t   Enter new description: ";
                getline(cin, task.description);
                cout << "\t\t\t\t\t   Enter new due date(yyyy-mm-dd): ";
                getline(cin, task.dueDate);
                cout << "\t\t\t\t\t   Task updated successfully!\n";
                return;
            }
        }
        cout << "\t\t\t\t\t   Task '" << taskName << "' not found.\n";
    }

    // Helper function to check if a task is due soon
    bool isDueSoon(string currentDate, string dueDate)
    {
        // Parse the current date
        int cYear, cMonth, cDay;
        sscanf(currentDate.c_str(), "%d-%d-%d", &cYear, &cMonth, &cDay);

        // Parse the due date
        int dYear, dMonth, dDay;
        sscanf(dueDate.c_str(), "%d-%d-%d", &dYear, &dMonth, &dDay);

        // Check if the due date is today or tomorrow
        if (cYear == dYear && cMonth == dMonth)
        {
            int daysLeft = dDay - cDay;
            return (daysLeft == 0 || daysLeft == 1); // Due today or tomorrow
        }

        return false;
    }

    // Function to display reminders for upcoming tasks
    void showReminders()
    {
        bool found = false;
        cout << "\t\t\t\t\t   Enter the current date (YYYY-MM-DD):";
        string currentDate;
        getline(cin, currentDate);

        cout << "\t\t\t\t\t   Upcoming tasks due soon:\n";
        for (auto &task : taskList)
        {
            if (!task.completed && isDueSoon(currentDate, task.dueDate))
            {
                task.displayTask();
                found = true;
            }
        }
        if (!found)
        {
            cout << "\t\t\t\t\t   No tasks are due within the next day.\n";
        }
    }
};

int main()
{
    ToDoListApp toDoListApp;
    int choice;
    string name, description, dueDate;

    while (choice != 7)
    {
        cout << "\t\t\t\t\t=============================================\n\t\t\t\t\t\t    To-Do List Application\n\t\t\t\t\t=============================================\n"
             << endl;
        cout << "\t\t\t\t\t   Add a new task___1" << endl;
        cout << "\t\t\t\t\t   Display all tasks__2" << endl;
        cout << "\t\t\t\t\t   Mark task as completed___3" << endl;
        cout << "\t\t\t\t\t   Delete a task__4" << endl;
        cout << "\t\t\t\t\t   Edit a task__5" << endl;
        cout << "\t\t\t\t\t   Show reminders___6" << endl;
        cout << "\t\t\t\t\t   For Exit___7" << endl;
        cout << "\t\t\t\t\t   Enter your choice: ";
        cin >> choice;
        cin.ignore(); // To clear the input buffer

        switch (choice)
        {
        case 1:
            system("cls");
            cout << "\t\t\t\t\t   Enter task name: ";
            getline(cin, name);
            cout << "\t\t\t\t\t   Enter task description: ";
            getline(cin, description);
            cout << "\t\t\t\t\t   Enter due date (YYYY-MM-DD): ";
            getline(cin, dueDate);

            toDoListApp.addTask(name, description, dueDate);
            getch();
            system("cls");
            break;

        case 2:
            system("cls");
            toDoListApp.displayTasks();
            getch();
            system("cls");
            break;

        case 3:
            system("cls");
            cout << "\t\t\t\t\t   Enter the name of the task to mark as completed: ";
            getline(cin, name);
            toDoListApp.markTaskCompleted(name);
            getch();
            system("cls");
            break;

        case 4:
            system("cls");
            cout << "\t\t\t\t\t   Enter the name of the task to delete: ";
            getline(cin, name);
            toDoListApp.deleteTask(name);
            getch();
            system("cls");
            break;

        case 5:
            system("cls");
            cout << "\t\t\t\t\t   Enter the name of the task to edit: ";
            getline(cin, name);
            toDoListApp.editTask(name);
            getch();
            system("cls");
            break;

        case 6:
            system("cls");
            toDoListApp.showReminders();
            getch();
            system("cls");
            break;

        case 7:
            system("cls");
            cout << "\t\t\t\t\t   Exiting application. Goodbye!\n";
            getch();
            system("cls");
            break;

        default:

            cout << "\t\t\t\t\t   Invalid choice. Please try again.\n";
            getch();
            system("cls");
            break;
        }
    }
    return 0;
}
